packages/hdwallet-keepkey/src/ethereum.ts

```diff
@@ -188,21 +173,116 @@ export async function ethGetAddress(transport: Transport, msg: core.ETHGetAddres
 }
 
 export async function ethSignMessage(transport: Transport, msg: core.ETHSignMessage): Promise<core.ETHSignedMessage> {
-  const m = new Messages.EthereumSignMessage();
+  const m = new Ethereum.EthereumSignMessage();
   m.setAddressNList(msg.addressNList);
   m.setMessage(ethers.utils.isBytes(msg.message) ? ethers.utils.arrayify(msg.message) : toUTF8Array(msg.message));
   const response = await transport.call(Messages.MessageType.MESSAGETYPE_ETHEREUMSIGNMESSAGE, m, {
     msgTimeout: core.LONG_TIMEOUT,
   });
-  const sig = response.proto as Messages.EthereumMessageSignature;
+  const sig = response.proto as Ethereum.EthereumMessageSignature;
   return {
     address: eip55.encode("0x" + core.toHexString(sig.getAddress_asU8())), // FIXME: this should be done in the firmware
     signature: "0x" + core.toHexString(sig.getSignature_asU8()),
   };
 }
 
+export async function ethSignTypedData(
+  transport: Transport,
+  msg: core.ETHSignTypedData
+): Promise<core.ETHSignedTypedData> {
+  /**
+   * If the message to be signed is sufficiently small, the KeepKey can calculate the
+   * domain separator and message hashes. Otherwise, we need to pre-calculate hashes
+   * here and verify on device.
+   */
+
+  let result: Ethereum.EthereumTypedDataSignature;
```

Comment:
not being reassigned within logic forks, may as well just assign as const locally scoped